### PR TITLE
test: Add api model entry for sgx

### PR DESCRIPTION
**Pull request by vtikoo**
**Reason for Change**:
Adds an api model entry for sgx, to enable e2e tests for sgx on AKS. 

**Issue Fixed**:

**Requirements**:

- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: test/e2e/test_cluster_configs/sgx.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+		"REGION_OPTIONS": "westeurope"
+	},
+	"apiModel": {
+        "apiVersion": "vlabs",
+        "properties": {
+          "orchestratorProfile": {
+            "orchestratorType": "Kubernetes",
+            "orchestratorRelease": "1.13",
+            "kubernetesConfig": {
+              "useManagedIdentity": true
```
Comment (jackfrancis) [inline test/e2e/test_cluster_configs/sgx.json]: Can we skip this part of the test? If not, more will be necessary to make MSI work w/ our E2E test runner. Here's the test output:

```
Error from deployment for kubernetes-westeurope-76947 in resource group kubernetes-westeurope-76947:exit status 1
 2019/10/04 00:16:00 Command Output: ERROR: Azure Error: InvalidTemplateDeployment
 Message: Deployment failed with multiple errors: 'Authorization failed for template resource '<resource>' of type 'Microsoft.Authorization/roleAssignments'. The client '<client>' with object id '<object>' does not have permission to perform action 'Microsoft.Authorization/roleAssignments/write' at scope '/subscriptions/<sub-id>/resourceGroups/kubernetes-westeurope-76947/providers/Microsoft.Authorization/roleAssignments/<role-assignment>'.:Authorization failed for template resource '<resource>' of type 'Microsoft.Authorization/roleAssignments'. The client '<client>' with object id '<object>' does not have permission to perform action 'Microsoft.Authorization/roleAssignments/write' at scope '/subscriptions/<sub-id>/resourceGroups/kubernetes-westeurope-76947/providers/Microsoft.Authorization/roleAssignments/<role-assignment>.'
```

Comment (vtikoo) [reply test/e2e/test_cluster_configs/sgx.json]: Sure, I dont think its essential for the e2e test.

**Comment by jackfrancis** [comment]
Is it possible the drivers don't work on 18.04-LTS? The api model in this PR fails thusly:

```
+ timeout 3600 apt-get -y install make gcc dkms
Reading package lists...
Building dependency tree...
Reading state information...
make is already the newest version (4.1-9.1ubuntu1).
dkms is already the newest version (2.3-3ubuntu9.5).
gcc is already the newest version (4:7.4.0-1ubuntu2.3).
0 upgraded, 0 newly installed, 0 to remove and 19 not upgraded.
+ break
+ echo Executed '"apt-get' -y install make gcc 'dkms"' 1 times
Executed "apt-get -y install make gcc dkms" 1 times
+ local SGX_DRIVER
++ basename https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin
+ SGX_DRIVER=sgx_linux_x64_driver_dcap_4f32b98.bin
+ local OE_DIR=/opt/azure/containers/oe
+ mkdir -p /opt/azure/containers/oe
+ retrycmd_if_failure 120 5 25 curl -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
+ retries=120
+ wait_sleep=5
+ timeout=25
+ shift
+ shift
+ shift
++ seq 1 120
+ for i in $(seq 1 $retries)
+ timeout 25 curl -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
+ break
+ echo Executed '"curl' -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o '/opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin"' 1 times
Executed "curl -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin" 1 times
+ chmod a+x /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
+ /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
Unpacking Intel SGX Driver ... done.
Verifying the integrity of the install package ... done.
Installing Intel SGX Driver ...
/tmp/sgx-driver-9U01QQ /var/lib/waagent/custom-script/download/1
install -d /opt/intel/sgxdriver/package
install -d /opt/intel/sgxdriver/scripts
cp -r package/* /opt/intel/sgxdriver/package
install scripts/* /opt/intel/sgxdriver/scripts
/var/lib/waagent/custom-script/download/1
/opt/intel/sgxdriver/package /var/lib/waagent/custom-script/download/1

Creating symlink /var/lib/dkms/sgx/0.10/source ->
                 /usr/src/sgx-0.10

DKMS: add completed.

Kernel preparation unnecessary for this kernel.  Skipping...

Building module:
cleaning build area...
'make' sign KDIR=/lib/modules/5.0.0-1020-azure/build....(bad exit status: 2)
ERROR (dkms apport): binary package for sgx: 0.10 not found
Error! Bad return status for module build on kernel: 5.0.0-1020-azure (x86_64)
Consult /var/lib/dkms/sgx/0.10/build/make.log for more information.
+ exit 91
```

**Comment by vtikoo** [comment]
We do support 18.04 afaik. Could you share the link of the failing Dockerfile or script?

**Comment by jackfrancis** [comment]
@vtikoo Does this api model work for you? Are you able to successfully build SGX clusters?

Specifically, what from the SGX drivers installation process would you like to see?

**Comment by vtikoo** [comment]
@jackfrancis we've had issues with missing linux headers, this might be an instance of the same? The workaround has been to manually apt install linux-headers with the correct version(which seems to be 5.0.0-1020-azure in this case).

**Comment by jackfrancis** [comment]
@vtikoo I tried that and it didn't help:

```
+ timeout 3600 apt-get -y install make gcc dkms linux-headers-5.0.0-1020-azure
Reading package lists...
Building dependency tree...
Reading state information...
make is already the newest version (4.1-9.1ubuntu1).
dkms is already the newest version (2.3-3ubuntu9.5).
gcc is already the newest version (4:7.4.0-1ubuntu2.3).
linux-headers-5.0.0-1020-azure is already the newest version (5.0.0-1020.21~18.04.1).
linux-headers-5.0.0-1020-azure set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 19 not upgraded.
+ break
+ echo Executed '"apt-get' -y install make gcc dkms 'linux-headers-5.0.0-1020-azure"' 1 times
Executed "apt-get -y install make gcc dkms linux-headers-5.0.0-1020-azure" 1 times
+ local SGX_DRIVER
++ basename https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin
+ SGX_DRIVER=sgx_linux_x64_driver_dcap_4f32b98.bin
+ local OE_DIR=/opt/azure/containers/oe
+ mkdir -p /opt/azure/containers/oe
+ retrycmd_if_failure 120 5 25 curl -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
+ retries=120
+ wait_sleep=5
+ timeout=25
+ shift
+ shift
+ shift
++ seq 1 120
+ for i in $(seq 1 $retries)
+ timeout 25 curl -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
+ break
+ echo Executed '"curl' -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o '/opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin"' 1 times
Executed "curl -fsSL https://download.01.org/intel-sgx/dcap-1.0.1/dcap_installer/ubuntuServer1804/sgx_linux_x64_driver_dcap_4f32b98.bin -o /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin" 1 times
+ chmod a+x /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
+ /opt/azure/containers/oe/sgx_linux_x64_driver_dcap_4f32b98.bin
Unpacking Intel SGX Driver ... done.
Verifying the integrity of the install package ... done.
Installing Intel SGX Driver ...
/tmp/sgx-driver-FfSXBm /var/lib/waagent/custom-script/download/1
install -d /opt/intel/sgxdriver/package
install -d /opt/intel/sgxdriver/scripts
cp -r package/* /opt/intel/sgxdriver/package
install scripts/* /opt/intel/sgxdriver/scripts
/var/lib/waagent/custom-script/download/1
/opt/intel/sgxdriver/package /var/lib/waagent/custom-script/download/1

Creating symlink /var/lib/dkms/sgx/0.10/source ->
                 /usr/src/sgx-0.10

DKMS: add completed.

Kernel preparation unnecessary for this kernel.  Skipping...

Building module:
cleaning build area...
'make' sign KDIR=/lib/modules/5.0.0-1020-azure/build....(bad exit status: 2)
ERROR (dkms apport): binary package for sgx: 0.10 not found
Error! Bad return status for module build on kernel: 5.0.0-1020-azure (x86_64)
Consult /var/lib/dkms/sgx/0.10/build/make.log for more information.
+ exit 91
```

Here is the `make.log` referenced above:

```
$ cat /var/lib/dkms/sgx/0.10/build/make.log
DKMS make.log for sgx-0.10 for kernel 5.0.0-1020-azure (x86_64)
Mon Oct  7 20:30:36 UTC 2019
make -C /lib/modules/5.0.0-1020-azure/build SUBDIRS=/var/lib/dkms/sgx/0.10/build CFLAGS_MODULE="-I/var/lib/dkms/sgx/0.10/build -I/var/lib/dkms/sgx/0.10/build/include" modules LE_ACTION=SIGN
make[1]: Entering directory '/usr/src/linux-headers-5.0.0-1020-azure'
Makefile:223: ================= WARNING ================
Makefile:224: 'SUBDIRS' will be removed after Linux 5.3
Makefile:225: Please use 'M=' or 'KBUILD_EXTMOD' instead
Makefile:226: ==========================================
  CC      /var/lib/dkms/sgx/0.10/build/le/main.o
  AS      /var/lib/dkms/sgx/0.10/build/le/entry.o
  CC      /var/lib/dkms/sgx/0.10/build/le/string.o
  LD      /var/lib/dkms/sgx/0.10/build/le/sgx_le_proxy
  AS [M]  /var/lib/dkms/sgx/0.10/build/sgx_le_proxy_piggy.o
  CC [M]  /var/lib/dkms/sgx/0.10/build/sgx_ioctl.o
  CC [M]  /var/lib/dkms/sgx/0.10/build/sgx_encl.o
/var/lib/dkms/sgx/0.10/build/sgx_encl.c: In function ‘sgx_process_add_page_req’:
/var/lib/dkms/sgx/0.10/build/sgx_encl.c:195:8: error: implicit declaration of function ‘vm_insert_pfn’; did you mean ‘vmf_insert_pfn’? [-Werror=implicit-function-declaration]
  ret = vm_insert_pfn(vma, addr, SGX_EPC_PFN(epc_page));
        ^~~~~~~~~~~~~
        vmf_insert_pfn
cc1: some warnings being treated as errors
scripts/Makefile.build:284: recipe for target '/var/lib/dkms/sgx/0.10/build/sgx_encl.o' failed
make[2]: *** [/var/lib/dkms/sgx/0.10/build/sgx_encl.o] Error 1
Makefile:1606: recipe for target '_module_/var/lib/dkms/sgx/0.10/build' failed
make[1]: *** [_module_/var/lib/dkms/sgx/0.10/build] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.0.0-1020-azure'
Makefile:96: recipe for target 'sign' failed
make: *** [sign] Error 2
```

**Comment by vtikoo** [comment]
@jackfrancis thanks for trying that out. One other thing I notice is that the driver url is different from what we recommend in our setup at https://github.com/openenclave/openenclave/blob/master/scripts/ansible/roles/linux/intel/vars/bionic.yml#L5.

**Comment by jackfrancis** [comment]
@vtikoo could you create a PR (you could do it in this PR) with corrections to the existing SGX drivers installation implementation? Sounds like the implementation in code right now is stale.

Thanks!

**Comment by vtikoo** [comment]
@jackfrancis have made the change, let me know if that fixes it.

**Comment by jackfrancis** [comment]
@vtikoo that did the trick, thanks. :)

btw I made a commit (to commit generated code updated as a result of your change)

Do we have any simple SGX-specific workloads that we can use to do Kubernetes application layer tests appropriate for this domain? As an example, this is a `Job` spec that we use to validate nvidia-backed node pools:

https://github.com/Azure/aks-engine/blob/master/test/e2e/kubernetes/workloads/cuda-vector-add.yaml

**Comment by vtikoo** [comment]
@jackfrancis awesome!
yeah we do have a simple Job, the image used is a private image though. 
```
apiVersion: batch/v1
kind: Job
metadata:
  name: sgxtest
spec:
  template:
    spec:
      containers:
      - name: sgxtest
        image: johnkord/sgxtest:0.1
        command: ["./sgxtesthost", "sgxtestenc.signed"]
        volumeMounts:
        - mountPath: /dev/sgx
          name: dev-sgx
        securityContext:
          privileged: true
      volumes:
      - name: dev-sgx
        hostPath:
          path: /dev/sgx
      restartPolicy: Never
  backoffLimit: 0
```

**Comment by vtikoo** [comment]
@jackfrancis pushed the above pasted Job which should run a simple hello world sgx application if sgx is set up properly, otherwise error out.

**Comment by jackfrancis** [comment]
@vtikoo I actually wired that job into our test runner

Is there any way we could use a more "official" job? Does the SGX team maintain the public dockerhub container image at `johnkord/sgxtest:0.1`?

File: test/e2e/test_cluster_configs/sgx.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            "vmSize": "Standard_D2s_v3",
+            "dnsPrefix": ""
+          },
+          "agentPoolProfiles": [
+            {
+              "name": "agentpool",
+              "count": 1,
+              "availabilityProfile": "VirtualMachineScaleSets",
+              "distro": "aks-ubuntu-18.04",
+              "vmSize": "Standard_DC4s",
```
Comment (jackfrancis) [inline test/e2e/test_cluster_configs/sgx.json]: @vtikoo in order to better manage quota on our test subs let's change this to `Standard_DC2s`. I imagine the hello world test job will be equivalently covered with a 2 core node, right?

**Comment by vtikoo** [comment]
@jackfrancis moved the docker image to the official oeciteam repository. The source dockerfile for the image is - https://github.com/openenclave/openenclave-aks/blob/master/k8s-sgxtest/Dockerfile

**Comment by vtikoo** [comment]
Thanks @jackfrancis for all the help!

**Pull request merged**